### PR TITLE
updated ceil, floor, round functions to use specific resolvers

### PR DESCRIPTION
**Pull request by chaudum**

### CODE REVIEW

File: sql/src/main/java/io/crate/operation/scalar/arithmetic/CeilFunction.java [hunk 1]
```diff
@@ -80,29 +86,24 @@ public Integer evaluate(Input[] args) {
...
-        }
+    private static class Resolver implements FunctionResolver {
 
         @Override
-        public Number evaluate(Input<Number>[] args) {
-            return args[0].value();
+        public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+            assert dataTypes.size() == 1 : "ceil function excepts exactly one argument type";
+            if (SINGLE_PRECISION.contains(dataTypes.get(0))) {
+                return new FloatCeilFunction(dataTypes);
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/arithmetic/CeilFunction.java]: would probably make sense to create static instances for every type?

File: sql/src/main/java/io/crate/operation/scalar/arithmetic/CeilFunction.java [hunk 2]
```diff
@@ -80,29 +86,24 @@ public Integer evaluate(Input[] args) {
...
-            info = new FunctionInfo(new FunctionIdent(NAME, ImmutableList.of(type)), type, FunctionInfo.Type.SCALAR,
-                FunctionInfo.DETERMINISTIC_AND_COMPARISON_REPLACEMENT);
-        }
+    private static class Resolver implements FunctionResolver {
 
         @Override
-        public Number evaluate(Input<Number>[] args) {
-            return args[0].value();
+        public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+            assert dataTypes.size() == 1 : "ceil function excepts exactly one argument type";
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/arithmetic/CeilFunction.java]: expects

**Comment by chaudum** [comment]
@dobe I made a static map with the function implementations. However, then we'd also need to register the UNDEFINED type. See fixup

File: sql/src/main/java/io/crate/operation/scalar/arithmetic/CeilFunction.java [hunk 3]
```diff
@@ -90,14 +93,20 @@ public Integer evaluate(Input[] args) {
...
         @Override
         public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
-            assert dataTypes.size() == 1 : "ceil function excepts exactly one argument type";
-            if (SINGLE_PRECISION.contains(dataTypes.get(0))) {
-                return new FloatCeilFunction(dataTypes);
-            } else {
-                return new DoubleCeilFunction(dataTypes);
-            }
+            assert dataTypes.size() == 1 : "ceil function expects exactly one argument type";
+            return INSTANCES.get(dataTypes.get(0));
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/arithmetic/CeilFunction.java]: need to throw IllegalArgumentException if type is not matching

File: sql/src/main/java/io/crate/operation/scalar/arithmetic/CeilFunction.java [hunk 4]
```diff
@@ -90,14 +93,20 @@ public Integer evaluate(Input[] args) {
 
     private static class Resolver implements FunctionResolver {
 
+        private static final Map<DataType, CeilFunction> INSTANCES = ImmutableMap.<DataType, CeilFunction>builder()
+            .put(DataTypes.FLOAT, new FloatCeilFunction(ImmutableList.of(DataTypes.FLOAT)))
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/arithmetic/CeilFunction.java]: pls start using Collections.singletonList

File: sql/src/main/java/io/crate/operation/scalar/arithmetic/RoundFunction.java [hunk 5]
```diff
@@ -88,14 +91,20 @@ public Number evaluate(Input[] args) {
...
+
         @Override
         public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
             assert dataTypes.size() == 1 : "round function excepts exactly one argument type";
-            if (SINGLE_PRECISION.contains(dataTypes.get(0))) {
-                return new FloatRoundFunction(dataTypes);
-            } else {
-                return new DoubleRoundFunction(dataTypes);
-            }
+            return INSTANCES.get(dataTypes.get(0));
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/arithmetic/RoundFunction.java]: also throw

**Pull request merged**